coffee/visitors/utilities.py

```diff
@@ -78,3 +81,106 @@ def visit_object(self, o, env):
 
     def visit_list(self, o, env):
         return [self.visit(e, env=env) for e in o]
+
+
+class Evaluate(Visitor):
+    """
+    Symbolically evaluate an expression enclosed in a loop nest, provided that
+    all of the symbols involved are constants and their value is known.
+
+    Return a dictionary mapping symbol names to (newly created) Decl nodes, each
+    declaration being initialized with a proper (newly computed and created)
+    ArrayInit object.
+
+    :arg decls: dictionary mapping symbol names to known Decl nodes.
+    """
+    def __init__(self, decls):
+        self.decls = decls
+        self.mapper = {
+            Sum: np.add,
+            Sub: np.subtract,
+            Prod: np.multiply,
+            Div: np.divide
+        }
+        super(Evaluate, self).__init__()
+
+    def visit_object(self, o, env):
+        return {}
+
+    def visit_list(self, o, env):
+        ret = {}
+        for entry in o:
+            ret.update(self.visit(entry, env=env))
+        return ret
+
+    def visit_Node(self, o, env):
+        ret = {}
+        for n in o.children:
+            ret.update(self.visit(n, env=env))
+        return ret
+
+    def visit_For(self, o, env):
+        try:
+            loop_nest = env["loop_nest"]
+        except KeyError:
+            loop_nest = []
+        new_env = Environment(env, loop_nest=loop_nest + [o])
+        return self.visit(o.body, env=new_env)
+
+    def visit_Writer(self, o, env):
+        try:
+            loop_nest = env["loop_nest"]
+        except KeyError:
+            loop_nest = []
+        lvalue = o.children[0]
+        writes = [l for l in loop_nest if l.dim in lvalue.rank]
+        # Evaluate the expression for each point in in the n-dimensional space
+        # represented by /writes/
+        dims = tuple(l.dim for l in writes)
+        shape = tuple(l.size for l in writes)
+        values = np.zeros(shape)
+        for i in itertools.product(*[range(j) for j in shape]):
+            point = {d: v for d, v in zip(dims, i)}
+            new_env = Environment(env, point=point)
+            # The sum takes into account reductions
+            values[i] = np.sum(self.visit(o.children[1], new_env))
+        return {lvalue: values}
+
+    def visit_BinExpr(self, o, env, *args, **kwargs):
+        if any([a is None for a in args]):
+            return
+        return self.mapper[o.__class__](*args)
+
+    def visit_Par(self, o, env):
+        return self.visit(o.child, env)
+
+    def visit_Symbol(self, o, env):
+        try:
+            point = env["point"]
+        except KeyError:
```

Comment:
No, this is somewhat "special", because it's not a top-level environment; rather, it's pushed (and successively popped) only by Writer nodes